querqy-for-lucene/querqy-lucene/src/main/java/querqy/lucene/rewrite/MultiMatchDismaxQueryStructurePostProcessor.java

```diff
@@ -0,0 +1,116 @@
+package querqy.lucene.rewrite;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * <p>A post-processor of a structure of {@link LuceneQueryFactory}s that wraps a
+ * {@link org.apache.lucene.search.DisjunctionMaxQuery} around clauses of top-level DisjunctionMaxQuerys that were
+ * created from the same input token.</p>
+ * <p>Let's assume an input query 'asus laptop' and a synonym expansion 'notebook' for laptop, and two query fields
+ * f1 and f2. This gives us DisMax(f1:asus | f2:asus) and DisMax(f1:laptop|f2:laptop|f1:notebook|f2:notebook) as DisMax
+ * queries, each wrapping the term queries that were derived from a common input token. A second level of grouping is
+ * applied per token value. This time we wrap a DisMax query around token values so that the disjunction-maximization of
+ * scores works between fields: DisMax(DisMax(f1:laptop|f2:laptop)|DisMax(f1:notebook|f2:notebook))</p>
+ * <p>Each grouping level comes with its own tie parameter, where the per-value grouping (dmqTieBreakerMultiplier)
+ * corresponds to the tie parameter as it is known from Lucene's dismax, and where the multiMatchTieBreakerMultiplier
+ * is the tie for the newly introduced grouping per input term:
+ * <pre>
+ *     DisMax(f1:asus | f2:asus)~dmqTieBreakerMultiplier
+ *     DisMax(
+ *          DisMax(f1:laptop|f2:laptop)~dmqTieBreakerMultiplier
+ *          |
+ *          DisMax(f1:notebook|f2:notebook)~dmqTieBreakerMultiplier
+ *     )~multiMatchTieBreakerMultiplier
+ * </pre>
+ *
+ */
+public class MultiMatchDismaxQueryStructurePostProcessor {
+
+    private final float dmqTieBreakerMultiplier;
+    private final float multiMatchTieBreakerMultiplier;
+
+    /**
+     *
+     * @param dmqTieBreakerMultiplier tiebreaker for per-value grouping
+     * @param multiMatchTieBreakerMultiplier - tiebreaker for grouping by input term
+     */
+    public MultiMatchDismaxQueryStructurePostProcessor(final float dmqTieBreakerMultiplier,
+                                                       final float multiMatchTieBreakerMultiplier) {
+        this.dmqTieBreakerMultiplier = dmqTieBreakerMultiplier;
+        this.multiMatchTieBreakerMultiplier = multiMatchTieBreakerMultiplier;
+    }
+
+    /**
+     * Rewrite the query structure. The operation might manipulate the input query structure and not operate on a copy.
+     *
+     * @param structure The input query structure.
+     * @return The rewritten structure.
+     */
+    public LuceneQueryFactory<?> process(final LuceneQueryFactory<?> structure) {
```

Comment:
What can I say... 

Some references to Uncle Bob:
- functions should not be large enough to hold nested control structures (conditionals and loops)
- blocks should be one line long, consisting probably of a single function call
- an ideal function is two to four lines of code long